File: backend/app/services/fuel_prices.py
"""Fuel price resolution.

Zimbabwe pump prices are capped monthly by ZERA (Zimbabwe Energy Regulatory
Authority). There is no public JSON API, so this module scrapes the published
figures and caches the last good result in the database.

If the network is unavailable or the page layout changes, the app degrades to
the last cached snapshot and marks it `is_live=False` so the UI can say so
honestly rather than presenting a stale number as current.
"""

import re
from datetime import datetime, timedelta, timezone

import httpx
from sqlalchemy import desc
from sqlalchemy.orm import Session

from ..models import PriceSnapshot

ZERA_URL = "https://www.zera.co.zw/fuel-pricing/"
CACHE_TTL = timedelta(hours=12)

# Used only on a completely cold database with no network. Clearly flagged as
# not live so the interface never claims it is a current figure.
COLD_START_PETROL = 1.57
COLD_START_DIESEL = 1.54

_PRICE_RE = re.compile(
    r"(blend|petrol|unleaded|diesel)[^0-9$]{0,80}\$?\s*([0-9]\.[0-9]{2,4})",
    re.IGNORECASE,
)


def _parse(html: str) -> tuple[float, float] | None:
    petrol: float | None = None
    diesel: float | None = None
    for label, value in _PRICE_RE.findall(html):
        price = float(value)
        if not 0.30 <= price <= 5.00:
            continue
        if label.lower() == "diesel" and diesel is None:
            diesel = price
        elif label.lower() != "diesel" and petrol is None:
            petrol = price
    if petrol is None or diesel is None:
        return None
    return petrol, diesel


def _latest(db: Session) -> PriceSnapshot | None:
    return db.query(PriceSnapshot).order_by(desc(PriceSnapshot.fetched_at)).first()


async def fetch_live_prices(db: Session, force: bool = False) -> PriceSnapshot:
    cached = _latest(db)
    if cached and not force:
        age = datetime.now(timezone.utc) - cached.fetched_at.replace(tzinfo=timezone.utc)
        if age < CACHE_TTL and cached.is_live:
            return cached

    try:
        async with httpx.AsyncClient(timeout=8.0, follow_redirects=True) as client:
            resp = await client.get(ZERA_URL, headers={"User-Agent": "FuelLink/1.0"})
            resp.raise_for_status()
        parsed = _parse(resp.text)
    except Exception:
        parsed = None

    if parsed:
        petrol, diesel = parsed
        snapshot = PriceSnapshot(
            petrol_price=petrol,
            diesel_price=diesel,
            source="ZERA fuel pricing page",
            source_url=ZERA_URL,
            is_live=True,
            effective_period=datetime.now(timezone.utc).strftime("%B %Y"),
        )
        db.add(snapshot)
        db.commit()
        db.refresh(snapshot)
        return snapshot

    if cached:
        return cached

    snapshot = PriceSnapshot(
        petrol_price=COLD_START_PETROL,
        diesel_price=COLD_START_DIESEL,
        source="Bundled fallback - not a live figure",
        source_url=ZERA_URL,
        is_live=False,
        effective_period="unverified",
    )
    db.add(snapshot)
    db.commit()
    db.refresh(snapshot)
    return snapshot


def current_unit_price(snapshot: PriceSnapshot, fuel_type: str | None) -> float:
    return snapshot.diesel_price if (fuel_type or "").lower() == "diesel" else snapshot.petrol_price

File: frontend/api/stations/fuel-prices.ts
import type { VercelRequest, VercelResponse } from "@vercel/node";
import { supabaseAdmin } from "../_lib/supabase.js";
import { handleError } from "../_lib/auth.js";

const ZERA_URL = "https://www.zera.co.zw/fuel-pricing/";
const CACHE_TTL_MS = 12 * 60 * 60 * 1000;
const COLD_PETROL = 1.57;
const COLD_DIESEL = 1.54;

const PRICE_RE = /(blend|petrol|unleaded|diesel)[^0-9$]{0,80}\$?\s*([0-9]\.[0-9]{2,4})/gi;

function parse(html: string): { petrol: number; diesel: number } | null {
  let petrol: number | null = null;
  let diesel: number | null = null;
  let m: RegExpExecArray | null;
  while ((m = PRICE_RE.exec(html))) {
    const price = parseFloat(m[2]);
    if (price < 0.30 || price > 5.00) continue;
    if (m[1].toLowerCase() === "diesel" && diesel === null) diesel = price;
    else if (m[1].toLowerCase() !== "diesel" && petrol === null) petrol = price;
  }
  if (petrol === null || diesel === null) return null;
  return { petrol, diesel };
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET,OPTIONS");
  if (req.method === "OPTIONS") return res.status(200).end();

  try {
    const force = req.query.refresh === "true";
    const { data: cached } = await supabaseAdmin
      .from("price_snapshots")
      .select("*")
      .order("fetched_at", { ascending: false })
      .limit(1)
      .maybeSingle();

    if (cached && !force) {
      const age = Date.now() - new Date(cached.fetched_at).getTime();
      if (age < CACHE_TTL_MS && cached.is_live) {
        return res.status(200).json(cached);
      }
    }

    let parsed: { petrol: number; diesel: number } | null = null;
    try {
      const resp = await fetch(ZERA_URL, {
        headers: { "User-Agent": "FuelLink/1.0" },
        signal: AbortSignal.timeout(8000),
      });
      const html = await resp.text();
      parsed = parse(html);
    } catch { /* network error */ }

    if (parsed) {
      const { data } = await supabaseAdmin
        .from("price_snapshots")
        .insert({
          petrol_price: parsed.petrol,
          diesel_price: parsed.diesel,
          source: "ZERA fuel pricing page",
          source_url: ZERA_URL,
          is_live: true,
          effective_period: new Date().toLocaleString("en-US", { month: "long", year: "numeric" }),
        })
        .select()
        .single();
      return res.status(200).json(data);
    }

    if (cached) return res.status(200).json(cached);

    const { data: fallback } = await supabaseAdmin
      .from("price_snapshots")
      .insert({
        petrol_price: COLD_PETROL,
        diesel_price: COLD_DIESEL,
        source: "Bundled fallback",
        source_url: ZERA_URL,
        is_live: false,
        effective_period: "unverified",
      })
      .select()
      .single();
    return res.status(200).json(fallback);
  } catch (err) {
    return handleError(res, err);
  }
}
